nixos/modules/system/activation/switch-to-configuration.pl

```diff
@@ -564,86 +619,58 @@ sub handle_modified_unit { ## no critic(Subroutines::ProhibitManyArgs, Subroutin
-
-            if (deepCmp($contentsOld->{$oldKey}, $contentsNew->{$newKey}) != 0) {
-                return (1, $contentsNew->{'X-ActivationStrategy'}[0] // "dynamic");
-            }
-        }
-    }
-
-    return (0, $contentsNew->{'X-ActivationStrategy'}[0] // "dynamic");
-}
-
-my $activeContainersOut = `machinectl list`;
-sub isContainerRunning {
-    my ($name) = @_;
-    if (index($activeContainersOut, $name) != -1) {
-        return 1;
-    }
-    return 0;
-}
-
-my @currentNspawnUnits = glob("/etc/systemd/nspawn/*.nspawn");
-my @upcomingNspawnUnits = glob("$out/etc/systemd/nspawn/*.nspawn");
-sub fingerprintUnit {
-    my ($s) = @_;
-    return abs_path($s) . (-f "${s}.d/overrides.conf" ? " " . abs_path "${s}.d/overrides.conf" : "");
-}
-foreach (@upcomingNspawnUnits) {
-    my $unit = basename($_);
-    $unit =~ s/\.nspawn//;
-    my $unitName = "systemd-nspawn\@$unit.service";
-    my $orig = $_;
-    $orig =~ s/^$out//;
-    if ($orig ~~ @currentNspawnUnits) {
-        if (fingerprintUnit($_) ne fingerprintUnit($orig)) {
-            my ($eq, $strategy) = compareNspawnUnits($orig, $_);
-            if ($strategy ne "none") {
-                if ($strategy ne "restart" and ($eq == 0 or $strategy eq "reload")) {
-                    if (isContainerRunning($unit) == 1) {
-                        $units_to_reload{$unitName} = 1;
-                    }
-                } elsif ($eq == 1) {
-                    $units_to_restart{$unitName} = 1;
-                }
+# Handle nspawn unit changes
+my @current_nspawn_units = glob("/etc/systemd/nspawn/*.nspawn");
+my @new_nspawn_units = glob("$out/etc/systemd/nspawn/*.nspawn");
+my %current_units_cmp = map { $_ => 1 } @current_nspawn_units;
+foreach my $new_unit_file (@new_nspawn_units) {
+    my $container_name = basename($new_unit_file);
+    $container_name =~ s/\.nspawn//;
+    my $unit_name = "systemd-nspawn\@$container_name.service";
+
+    my $cur_unit_file = $new_unit_file;
+    $cur_unit_file =~ s/^$out//;
+    if (exists($current_units_cmp{$cur_unit_file})) {
+        my %new_unit_info = parse_unit($new_unit_file);
+        my $strategy = $new_unit_info{"Exec"}{"X-ActivationStrategy"}[0] // "dynamic";
+
+        # Skip comparison logic/restart check if ActivationStrategy is "none"
+        next if $strategy eq "none";
+
+        my %cur_unit_info = parse_unit($cur_unit_file);
+        my $changed = compare_nspawn_units(\%cur_unit_info, \%new_unit_info);
+
+=pod Truth table for restarts
+|Strategy|Changed|Reload|Restart|
+|--------|-------|------|-------|
+|Dynamic |0      |Y     |-      |
+|Dynamic |1      |-     |Y      |
+|Restart |0      |-     |-      |
+|Restart |1      |-     |Y      |
+|Reload  |0      |Y     |-      |
+|Reload  |1      |Y     |-      |
+=cut
+        if ($strategy ne "restart" and ($changed == 0 or $strategy eq "reload")) {
+            if (is_container_running($container_name) == 1) {
+                $units_to_reload{$unit_name} = 1;
```

Comment:
I'll need to re-read this part tomorrow and run the tests afterwords since I'm pretty tired right now.... can't promise that because heavy exam phase, but this weekend I'd merge this into my branch and try to "un-draft" my PR against upstream

Reply:
No worries! I hope your exams are going well, and certainly they take priority over this. 😄

Reply:
Exams survived, will try to resume now.